### PR TITLE
PS-5910: Support V3 encryption header

### PR DESCRIPTION
**Pull request by dutow**
This fixes some 32-64 bit differences, and the 5.7->8.0
redo log encryption upgrade encryption issue.

### CODE REVIEW

File: storage/innobase/fsp/fsp0fsp.cc [hunk 1]
```diff
@@ -978,17 +978,23 @@ fsp_header_fill_encryption_info(
...
 		memcpy(ptr, ENCRYPTION_KEY_MAGIC_V2, ENCRYPTION_MAGIC_SIZE);
+	} else {
+		memcpy(ptr, ENCRYPTION_KEY_MAGIC_V3, ENCRYPTION_MAGIC_SIZE);
 	}
 	ptr += ENCRYPTION_MAGIC_SIZE;
 
 	/* Write master key id. */
 	mach_write_to_4(ptr, master_key_id);
-	ptr += sizeof(ulint);
+	if (version == Encryption::ENCRYPTION_VERSION_3) {
```
Comment (satya-bodapati) [inline storage/innobase/fsp/fsp0fsp.cc:990]: While writing, there is no need to write V1 or V2. Just write V3 format (No need check for 'version' during write)

Change the entire patch to
1. Write only V3
2. Handle reading of v1, v2 and v3.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/fsp/fsp0fsp.cc (hunk 1 above)
Comment (dutow) [reply storage/innobase/fsp/fsp0fsp.cc:990]: I have to keep at least V1: some encryption features are already GA, and V1 stored keys in the keyring under a different name.
For V2, we would lose some (not 100% complete) downgrade ability within 5.7, that's probably not an issue - but then it's only a few lines of code.

Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:990]: So tablespaces that come from older 5.7 continue to use V2?  I don't think we should care about downgrade ability before GA.

Also ALTER INSTANCE INNODB ROTATE MASTER KEY should write encryption headers in V3. Please verify this.

Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:990]: any rebuild of the tables would bring the tables to V3 format anyway

Comment (dutow) [reply storage/innobase/fsp/fsp0fsp.cc:990]: This PR is for 5.7, it is not about the 5.7->8.0 upgrade, it's about how versions within 5.7 behave. And some encryption features are already GA on 5.7.

Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:990]: @dutow I understand that. So my questions are
1. ALTERs which rebuild table would recreate the tablespace encryption header in V3 format right? please check and confirm
2. ALTER INSTANCE .. on upgraded tables from earlier version. Would they continue using V2 format after rotation?

Comment (dutow) [reply storage/innobase/fsp/fsp0fsp.cc:990]: I've added a testcase about this.

File: mysql-test/suite/innodb/t/table_encrypt_upgrade-master.opt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+--early-plugin-load="keyring_file=$KEYRING_PLUGIN"
```
Comment (satya-bodapati) [inline mysql-test/suite/innodb/t/table_encrypt_upgrade-master.opt]: use $KEYRING_PLUGIN_EARLY_LOAD

File: mysql-test/suite/innodb/t/table_encrypt_upgrade.test [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+let SEARCH_PATTERN=lCB;
+--source include/search_pattern.inc
+LET $MYSQLD_DATADIR = `select @@datadir`;
+let SEARCH_PATTERN=lCC;
+--source include/search_pattern.inc
+let SEARCH_PATTERN=znmcg;
+--source include/search_pattern.inc
+
+# Rebuild the table
+ALTER TABLE t1 ENGINE = InnoDB ENCRYPTION="Y";
```
Comment (satya-bodapati) [inline mysql-test/suite/innodb/t/table_encrypt_upgrade.test:57]: Add one more V2 TABLE. After restart, do ROTATE INNODB MASTER KEY.

check if the IBD has v2 or v3?

File: storage/innobase/fsp/fsp0fsp.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:990]: Thanks, looks good. added comment on that

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: mysql-test/suite/innodb/t/table_encrypt_upgrade.test (hunk 3 above)
Comment (dutow) [reply mysql-test/suite/innodb/t/table_encrypt_upgrade.test:57]: Done. Also removed the unnecessary restarts.

File: storage/innobase/fsp/fsp0fsp.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:990]: Now your tests confirm that DDLs (and ALTER INSTANCE) will change all tables to V3 and downgrade is not possible.

I will make this suggestion optional. So upto you if you want to write only V3.

**Review by satya-bodapati — APPROVED** [review]

**Comment by satya-bodapati** [comment]
btw, can you bring  the import export test from 8.0? f565013b7f31f225c349a6484ccab8cf4b5c8f52

I've reviewed this patch too at Oracle :)

File: storage/innobase/fsp/fsp0fsp.cc (hunk 1 above)
Comment (dutow) [reply storage/innobase/fsp/fsp0fsp.cc:990]: I'll keep this functionality for now - without it, I wouldn't be able to add this test :)

**Pull request merged**